aes.c

```diff
@@ -570,3 +574,69 @@ void AES_CTR_xcrypt_buffer(struct AES_ctx* ctx, uint8_t* buf, size_t length)
 
 #endif // #if defined(CTR) && (CTR == 1)
 
+
+#if defined(CMAC) && (CMAC == 1)
+
+#define AES_CMAC_MSB 0x80
+#define AES_CMAC_RB 0x87
+
+static void GenerateSubkey(uint8_t* key)
+{
+  uint8_t msb = key[0] & AES_CMAC_MSB;
+  uint8_t i;
+
+  for (i = 0; i < AES_BLOCKLEN - 1; ++i)
+  {
+    key[i] = (key[i] << 1) | (key[i + 1] >> 7);
+  }
+  key[AES_BLOCKLEN - 1] <<= 1;
+
+  if (msb)
+  {
+    key[AES_BLOCKLEN - 1] ^= AES_CMAC_RB;
+  }
+}
+
+void AES_CMAC_generate(const struct AES_ctx* ctx, const uint8_t* buf, size_t length, uint8_t* buf_out, size_t length_out)
+{
+  uint8_t block[AES_BLOCKLEN] = {0};
+  uint8_t L[AES_BLOCKLEN] = {0};
+  size_t i;
+
+  /* First subkey. */
+  Cipher((state_t*)L, ctx->RoundKey);
+  GenerateSubkey(L);
+
+  do
+  {
+    if (length < AES_BLOCKLEN)
+    {
+      /* Last block is not complete. Padding and using second subkey. */
+      GenerateSubkey(L);
+      L[length] ^= AES_CMAC_MSB;
+    }
+    if (length <= AES_BLOCKLEN)
+    {
+      for (i = 0; i < length; ++i)
+      {
+        L[i] ^= buf[i];
+      }
+      length = AES_BLOCKLEN;
+      buf = L;
+    }
+    Xor(block, buf);
+    Cipher((state_t*)block, ctx->RoundKey);
+    length -= AES_BLOCKLEN;
+    buf += AES_BLOCKLEN;
+  }
+  while (length);
+
+  for (i = 0; i < sizeof(block) && i < length_out; ++i)
```

Comment:
I'd like an extra set of parentheses here, to avoid confusion about operator precedence.
That way, it becomes like this:

```c
for (i = 0; (i < sizeof(block)) && (i < length_out); ++i)
```